File: java/sdk/sdk-core/src/main/java/com/novax/sdk/core/http/interceptors/LoggingInterceptor.java
package com.novax.sdk.core.http.interceptors;

import com.novax.sdk.core.http.Interceptor;
import com.novax.sdk.core.http.SdkRequest;
import com.novax.sdk.core.http.SdkResponse;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * Minimal stdout/stderr-style request logger. Replace with SLF4J/Logback at will.
 *
 * <p>Normal responses log at {@code INFO} (status + elapsed). Server-error
 * responses ({@code statusCode >= 500}) log at {@code SEVERE} and include the
 * response body (truncated to {@link #ERROR_BODY_MAX_CHARS} chars) so a 500 in
 * production leaves enough breadcrumbs to diagnose.
 */
public final class LoggingInterceptor implements Interceptor {

    private static final Logger LOG = Logger.getLogger(LoggingInterceptor.class.getName());
    private static final int ERROR_BODY_MAX_CHARS = 2048;

    @Override
    public SdkResponse intercept(Chain chain) throws IOException {
        SdkRequest req = chain.request();
        long t0 = System.nanoTime();
        LOG.log(Level.INFO, () -> "-> " + req.method() + " " + req.uri());
        SdkResponse resp = chain.proceed(req);
        long ms = (System.nanoTime() - t0) / 1_000_000;

        if (resp.statusCode() >= 500) {
            LOG.log(Level.SEVERE, () -> "<- " + resp.statusCode() + " (" + ms + "ms) "
                    + req.method() + " " + req.uri()
                    + " body=" + truncate(bodyAsString(resp), ERROR_BODY_MAX_CHARS));
        } else {
            LOG.log(Level.INFO, () -> "<- " + resp.statusCode() + " (" + ms + "ms)");
        }
        return resp;
    }

    private static String bodyAsString(SdkResponse resp) {
        byte[] body = resp.body();
        if (body == null || body.length == 0) return "<empty>";
        return new String(body, StandardCharsets.UTF_8);
    }

    private static String truncate(String s, int max) {
        if (s.length() <= max) return s;
        return s.substring(0, max) + "…(+" + (s.length() - max) + " chars)";
    }
}

File: python/novax_sdk/http/interceptors/logging_interceptor.py
import logging
import time

from novax_sdk.http.interceptor import Chain, Interceptor
from novax_sdk.http.sdk_response import SdkResponse

_LOG = logging.getLogger(__name__)
_ERROR_BODY_MAX = 2048


class LoggingInterceptor(Interceptor):
    """Logs request/response. INFO normally; ERROR on HTTP >=500 with body."""

    def intercept(self, chain: Chain) -> SdkResponse:
        req = chain.request
        _LOG.info("-> %s %s", req.method.value, req.url)
        t0 = time.monotonic()
        resp = chain.proceed(req)
        ms = int((time.monotonic() - t0) * 1000)
        if resp.status_code >= 500:
            body_str = resp.body.decode(errors="replace") if resp.body else "<empty>"
            if len(body_str) > _ERROR_BODY_MAX:
                body_str = body_str[:_ERROR_BODY_MAX] + f"…(+{len(body_str)-_ERROR_BODY_MAX} chars)"
            _LOG.error("<- %s (%dms) %s %s body=%s",
                       resp.status_code, ms, req.method.value, req.url, body_str)
        else:
            _LOG.info("<- %s (%dms)", resp.status_code, ms)
        return resp
